File: Examples/DemoApp/iosApp/DemoApp/UI/ObservableStore.swift
//
//  ObservableStore.swift
//  DemoApp
//
//  Created by Jérémy TOUZY on 22/01/2020.
//  Copyright © 2020 jtouzy. All rights reserved.
//

import shared

class ObservableStore<T: ViewState>: Store, ObservableObject {
    var currentState: ViewState {
        didSet {
            guard let newState = viewState as? T else {
                fatalError("Cannot be casted")
            }
            state = newState
        }
    }
    @Published var state: T

    init(baseState: T) {
        viewState = baseState
        state = baseState
    }
}

File: Examples/DemoApp/androidApp/src/main/kotlin/com/jtouzy/demo/app/ui/ObservableStore.kt
package com.jtouzy.demo.app.ui

import androidx.compose.Model
import com.jtouzy.demo.ui.Store
import com.jtouzy.demo.ui.ViewState

@Model
class ObservableStore<T: ViewState>(override var currentState: T) : Store<T> {
}
